x/wasm/keeper/keeper.go

```diff
@@ -1005,6 +1029,116 @@ func (k Keeper) newQueryHandler(ctx sdk.Context, contractAddress sdk.AccAddress)
 	return NewQueryHandler(ctx, k.wasmVMQueryHandler, contractAddress, k.gasRegister)
 }
 
+func (k Keeper) updateSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress, sizeChange int64) error {
+	// always charge rent first before updating size, as the charging logic assumes constant size since last
+	// charged block
+	if err := k.chargeRent(ctx, contractAddress); err != nil {
+		return err
+	}
+	sizeKey := types.GetContractStoreSizePrefix(contractAddress)
+	store := ctx.KVStore(k.storeKey)
+	currentSize, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return types.ErrUpdateContractSize.Wrap(err.Error())
+	}
+	currentSizeSigned := int64(currentSize) + sizeChange
+	if currentSizeSigned < 0 {
+		// this should not happen
+		panic("negative size for contract")
+	}
+	newBz := make([]byte, 8)
+	binary.BigEndian.PutUint64(newBz, uint64(currentSizeSigned))
+	store.Set(sizeKey, newBz)
+	return nil
+}
+
+func (k Keeper) getCurrentSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress) (uint64, error) {
+	return GetSize(ctx, contractAddress, k.storeKey)
+}
+
+func (k Keeper) chargeRent(ctx sdk.Context, contractAddress sdk.AccAddress) error {
+	rentInfo, err := k.getRentInfo(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	duration := uint64(ctx.BlockHeight()) - rentInfo.LastChargedBlock
+	if duration == 0 {
+		// shortcut in the case of same block changes
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	size, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	if size == 0 {
+		rentInfo.LastChargedBlock = uint64(ctx.BlockHeight())
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	rentPrice := k.GetParams(ctx).UnitRentPrice
+	rentToBeCharged := sdk.NewDec(int64(duration * size)).Mul(rentPrice).RoundInt().Int64()
+	// protect against overflow
+	if rentInfo.Balance < rentToBeCharged+math.MinInt64 {
+		rentInfo.Balance = math.MinInt64
+	} else {
+		rentInfo.Balance -= rentToBeCharged
+	}
+	rentInfo.LastChargedBlock = uint64(ctx.BlockHeight())
+	k.setRentInfo(ctx, contractAddress, rentInfo)
```

Comment:
should we put this after the error check for slight optimization (then the write doesnt have to occur to just rollback if the balance is negative)

Reply:
hm good point i'm doing it wrong here now that you mentioned rollback. My intention was to persist the negative balance so that the contract owner would still be responsible for the accrued rent for his state between when the contract is "frozen" and when he deposits and thaws it, but forgot it wouldn't persist due to the rollback. We might need to do something like:
```
DeliverTx -> branch cachekv
  WASM execution -> branch of branch of cachekv
  Charge rent
     -> if success, commits branch of branch
     -> if out of rent, revert branch of branch
DeliverTX -> branch cachekv commits/aborts regardless of out of rent or not
```

Reply:
actually a better way would be to place the rent deduction logic into ante handler and the check in msgServer, so that we leverage the existing rollback infra instead of throwing in yet another layer